ghost/core/core/server/services/link-tracking/PostLinkRepository.js

```diff
@@ -48,6 +62,116 @@ module.exports = class PostLinkRepository {
         return result;
     }
 
+    /**
+     * validate and manage the new redirect url in filter
+     * `to` url needs decoding and transformation to relative url for comparision
+     * @param {string} filter
+     * @returns {Object} parsed filter
+     * @throws {errors.BadRequestError}
+     */
+    #parseLinkFilter(filter) {
+        // decode filter to manage any encoded uri components
+        filter = decodeURIComponent(filter);
+
+        try {
+            const filterJson = nql(filter).parse();
+            const postId = filterJson?.$and?.[0]?.post_id;
+            const redirectUrl = new URL(filterJson?.$and?.[1]?.to);
+            if (!postId || !redirectUrl) {
+                throw new errors.BadRequestError({
+                    message: tpl(messages.invalidFilter)
+                });
+            }
+            return {
+                postId,
+                redirectUrl
+            };
+        } catch (e) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.invalidFilter),
+                context: e.message
+            });
+        }
+    }
+
+    #getRedirectLinkWithAttribution({newLink, oldLink, postId}) {
+        const newUrl = new URL(newLink);
+        const oldUrl = new URL(oldLink);
+        // append newsletter ref query param from oldUrl to newUrl
+        if (oldUrl.searchParams.has('ref')) {
+            newUrl.searchParams.set('ref', oldUrl.searchParams.get('ref'));
+        }
+
+        // append post attribution to site urls
+        const isSite = this.#urlUtils.isSiteUrl(newUrl);
+        if (isSite) {
+            newUrl.searchParams.set('attribution_type', 'post');
+            newUrl.searchParams.set('attribution_id', postId);
+        }
+        return newUrl;
+    }
+
+    async updateLinks(data, options) {
```

Comment:
Same as previous comment, I think there's too much application logic in this class, and it should be pulled out to the tracking/redirect package.

This class can probably just be updated to something like:
```
/**
 * @param {PostLink[]} postLinks
 * @returns {Promise<smth>}
 */ 
async saveMany(postLinks) {
  // impl
}
```

With the fetching and mapping of values happening outside of the repository

Reply:
Updated to move logic out to `LinkTrackingService` and kept this to just update given linkIds with new data.